### PR TITLE
Repeat latched messages at start of bag

### PR DESCRIPTION
**Pull request by tappan-at-git**
Adds a --repeat-latched option to `rosbag record`. This makes `Recorder` maintain a {(topic, publisher):  message} map, then iterate over that map for each new bagfile. In theory this could cause problems if the total size of latched messages exceeds the max size of the bag, if that seems like a reasonable cause for concern I can skip writing latched messages if the total size of the map gets too large.

### CODE REVIEW

File: tools/rosbag/src/record.cpp [hunk 1]
```diff
@@ -227,7 +230,7 @@ rosbag::RecorderOptions parseOptions(int argc, char** argv) {
       else
         throw ros::Exception("Duration unit must be s, m, or h");
 
-      
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/record.cpp]:
```suggestion
      
```

File: tools/rosbag/src/record.cpp [hunk 2]
```diff
@@ -296,6 +299,6 @@ int main(int argc, char** argv) {
     // Run the recorder
     rosbag::Recorder recorder(opts);
     int result = recorder.run();
-    
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/record.cpp]:
```suggestion
    
```

File: tools/rosbag/src/recorder.cpp [hunk 3]
```diff
@@ -251,7 +251,7 @@ bool Recorder::shouldSubscribeToTopic(std::string const& topic, bool from_node)
     if(options_.record_all || from_node) {
         return true;
     }
-    
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/recorder.cpp]:
```suggestion
    
```

File: tools/rosbag/src/recorder.cpp [hunk 4]
```diff
@@ -285,18 +285,31 @@ std::string Recorder::timeToStr(T ros_t)
 void Recorder::doQueue(const ros::MessageEvent<topic_tools::ShapeShifter const>& msg_event, string const& topic, shared_ptr<ros::Subscriber> subscriber, shared_ptr<int> count) {
     //void Recorder::doQueue(topic_tools::ShapeShifter::ConstPtr msg, string const& topic, shared_ptr<ros::Subscriber> subscriber, shared_ptr<int> count) {
     Time rectime = Time::now();
-    
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/recorder.cpp]:
```suggestion
    
```

File: tools/rosbag/src/recorder.cpp [hunk 5]
```diff
@@ -285,18 +285,31 @@ std::string Recorder::timeToStr(T ros_t)
...
     if (options_.verbose)
         cout << "Received message on topic " << subscriber->getTopic() << endl;
 
     OutgoingMessage out(topic, msg_event.getMessage(), msg_event.getConnectionHeaderPtr(), rectime);
-    
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/recorder.cpp]:
```suggestion
    
```

File: tools/rosbag/src/recorder.cpp [hunk 6]
```diff
@@ -312,7 +325,7 @@ void Recorder::doQueue(const ros::MessageEvent<topic_tools::ShapeShifter const>&
             }
         }
     }
-  
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/recorder.cpp]:
```suggestion
  
```

File: tools/rosbag/src/recorder.cpp [hunk 7]
```diff
@@ -365,9 +378,9 @@ void Recorder::updateFilenames() {
 void Recorder::snapshotTrigger(std_msgs::Empty::ConstPtr trigger) {
     (void)trigger;
     updateFilenames();
-    
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/recorder.cpp]:
```suggestion
    
```

File: tools/rosbag/src/recorder.cpp [hunk 8]
```diff
@@ -365,9 +378,9 @@ void Recorder::updateFilenames() {
...
     ROS_INFO("Triggered snapshot recording with name %s.", target_filename_.c_str());
-    
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/recorder.cpp]:
```suggestion
    
```

File: tools/rosbag/src/recorder.cpp [hunk 9]
```diff
@@ -527,9 +549,9 @@ void Recorder::doRecord() {
         OutgoingMessage out = queue_->front();
         queue_->pop();
         queue_size_ -= out.msg->size();
-        
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/recorder.cpp]:
```suggestion
        
```

File: tools/rosbag/src/recorder.cpp [hunk 10]
```diff
@@ -527,9 +549,9 @@ void Recorder::doRecord() {
...
         lock.release()->unlock();
-        
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/recorder.cpp]:
```suggestion
        
```

File: tools/rosbag/src/recorder.cpp [hunk 11]
```diff
@@ -554,23 +576,23 @@ void Recorder::doRecord() {
 
 void Recorder::doRecordSnapshotter() {
     ros::NodeHandle nh;
-  
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/recorder.cpp]:
```suggestion
  
```

File: tools/rosbag/src/recorder.cpp [hunk 12]
```diff
@@ -554,23 +576,23 @@ void Recorder::doRecord() {
...
     while (nh.ok() || !queue_queue_.empty()) {
         boost::unique_lock<boost::mutex> lock(queue_mutex_);
         while (queue_queue_.empty()) {
             if (!nh.ok())
                 return;
             queue_condition_.wait(lock);
         }
-        
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/recorder.cpp]:
```suggestion
        
```

File: tools/rosbag/src/recorder.cpp [hunk 13]
```diff
@@ -554,23 +576,23 @@ void Recorder::doRecord() {
...
         OutgoingQueue out_queue = queue_queue_.front();
         queue_queue_.pop();
-        
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/recorder.cpp]:
```suggestion
        
```

File: tools/rosbag/src/recorder.cpp [hunk 14]
```diff
@@ -554,23 +576,23 @@ void Recorder::doRecord() {
...
         lock.release()->unlock();
-        
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/recorder.cpp]:
```suggestion
        
```

File: tools/rosbag/src/recorder.cpp [hunk 15]
```diff
@@ -554,23 +576,23 @@ void Recorder::doRecord() {
...
         string target_filename = out_queue.filename;
         string write_filename  = target_filename + string(".active");
-        
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/recorder.cpp]:
```suggestion
        
```

File: tools/rosbag/src/recorder.cpp [hunk 16]
```diff
@@ -600,7 +622,7 @@ void Recorder::doCheckMaster(ros::TimerEvent const& e, ros::NodeHandle& node_han
 	        subscribe(t.name);
 	}
     }
-    
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/recorder.cpp]:
```suggestion
    
```

File: tools/rosbag/src/recorder.cpp [hunk 17]
```diff
@@ -625,7 +647,7 @@ void Recorder::doCheckMaster(ros::TimerEvent const& e, ros::NodeHandle& node_han
           XmlRpc::XmlRpcValue resp2;
           req2[0] = ros::this_node::getName();
           c.execute("getSubscriptions", req2, resp2);
-          
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/recorder.cpp]:
```suggestion
          
```

File: tools/rosbag/src/recorder.cpp [hunk 18]
```diff
@@ -692,8 +714,8 @@ bool Recorder::checkDisk() {
     {
         info = boost::filesystem::space(p);
     }
-    catch (boost::filesystem::filesystem_error &e) 
-    { 
+    catch (boost::filesystem::filesystem_error &e)
+    {
```
Comment (tappan-at-git) [inline tools/rosbag/src/recorder.cpp]:
```suggestion
    { 
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 19]
```diff
@@ -187,7 +189,7 @@ def info_cmd(argv):
             b.close()
             if i < len(args) - 1:
                 print('---')
-        
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
        
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 20]
```diff
@@ -353,7 +355,7 @@ def expr_eval(expr):
     filter_fn = expr_eval(expr)
 
     outbag = Bag(outbag_filename, 'w')
-    
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
    
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 21]
```diff
@@ -366,10 +368,10 @@ def expr_eval(expr):
     try:
         meter = ProgressMeter(outbag_filename, inbag._uncompressed_size)
         total_bytes = 0
-    
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
    
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 22]
```diff
@@ -366,10 +368,10 @@ def expr_eval(expr):
...
         if options.verbose_pattern:
             verbose_pattern = expr_eval(options.verbose_pattern)
-    
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
    
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 23]
```diff
@@ -394,7 +396,7 @@ def expr_eval(expr):
 
                 total_bytes += len(serialized_bytes)
                 meter.step(total_bytes)
-        
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
        
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 24]
```diff
@@ -415,7 +417,7 @@ def fix_cmd(argv):
 
     inbag_filename  = args[0]
     outbag_filename = args[1]
-    rules           = args[2:]   
+    rules           = args[2:]
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
    rules           = args[2:]   
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 25]
```diff
@@ -453,7 +455,7 @@ def fix_cmd(argv):
         options.noplugins = False
 
     migrator = MessageMigrator(rules, plugins=not options.noplugins)
-    
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
    
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 26]
```diff
@@ -471,12 +473,12 @@ def fix_cmd(argv):
         print('Bag could not be migrated.  The following migrations could not be performed:')
         for m in migrations:
             print_trans(m[0][0].old_class, m[0][-1].new_class, 0)
-            
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
            
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 27]
```diff
@@ -471,12 +473,12 @@ def fix_cmd(argv):
...
             if len(m[1]) > 0:
                 print('    %d rules missing:' % len(m[1]))
                 for r in m[1]:
                     print_trans(r.old_class, r.new_class,1)
-                    
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
                    
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 28]
```diff
@@ -498,13 +500,13 @@ def check_cmd(argv):
             parser.error('The file %s already exists.  Include -a if you intend to append.' % options.rulefile)
         if not rulefile_exists and options.append:
             parser.error('The file %s does not exist, and so -a is invalid.' % options.rulefile)
-    
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
    
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 29]
```diff
@@ -498,13 +500,13 @@ def check_cmd(argv):
...
     if options.append:
         append_rule = [options.rulefile]
     else:
         append_rule = []
 
-    # First check that the bag is not unindexed 
+    # First check that the bag is not unindexed
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
    # First check that the bag is not unindexed 
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 30]
```diff
@@ -517,11 +519,11 @@ def check_cmd(argv):
     mm = MessageMigrator(args[1:] + append_rule, not options.noplugins)
 
     migrations = checkbag(mm, args[0])
-       
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py:522]:
```suggestion
       
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 31]
```diff
@@ -517,11 +519,11 @@ def check_cmd(argv):
...
     if len(migrations) == 0:
         print('Bag file does not need any migrations.')
         exit(0)
-        
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
        
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 32]
```diff
@@ -604,9 +606,9 @@ def decompress_cmd(argv):
 
     if len(args) < 1:
         parser.error('You must specify at least one bag file.')
-    
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
    
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 33]
```diff
@@ -604,9 +606,9 @@ def decompress_cmd(argv):
...
     op = lambda inbag, outbag, quiet: change_compression_op(inbag, outbag, Compression.NONE, options.quiet)
-    
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
    
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 34]
```diff
@@ -620,7 +622,7 @@ def reindex_cmd(argv):
 
     if len(args) < 1:
         parser.error('You must specify at least one bag file.')
-    
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
    
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 35]
```diff
@@ -691,12 +693,12 @@ def bag_op(inbag_filenames, allow_unindexed, open_inbag, copy_fn, op, output_dir
         if outbag_filename == inbag_filename:
             # Rename the input bag to ###.orig.###, and open for reading
             backup_filename = '%s.orig%s' % os.path.splitext(inbag_filename)
-            
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
            
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 36]
```diff
@@ -691,12 +693,12 @@ def bag_op(inbag_filenames, allow_unindexed, open_inbag, copy_fn, op, output_dir
...
             if not force and os.path.exists(backup_filename):
                 if not quiet:
                     print('Skipping %s. Backup path %s already exists.' % (inbag_filename, backup_filename), file=sys.stderr)
                 continue
-            
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
            
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 37]
```diff
@@ -705,7 +707,7 @@ def bag_op(inbag_filenames, allow_unindexed, open_inbag, copy_fn, op, output_dir
             except OSError as ex:
                 print('ERROR %s %s to %s: %s' % ('copying' if copy else 'moving', inbag_filename, backup_filename, str(ex)), file=sys.stderr)
                 continue
-            
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
            
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 38]
```diff
@@ -771,7 +773,7 @@ def bag_op(inbag_filenames, allow_unindexed, open_inbag, copy_fn, op, output_dir
                 except OSError as ex:
                     print('ERROR %s %s to %s: %s', ('removing' if copy else 'moving', backup_filename, inbag_filename, str(ex)), file=sys.stderr)
                     break
-    
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
    
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 39]
```diff
@@ -787,12 +789,12 @@ def change_compression_op(inbag, outbag, compression, quiet):
         total_bytes = 0
         for topic, msg, t, conn_header in inbag.read_messages(raw=True, return_connection_header=True):
             msg_type, serialized_bytes, md5sum, pos, pytype = msg
-    
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
    
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 40]
```diff
@@ -787,12 +789,12 @@ def change_compression_op(inbag, outbag, compression, quiet):
...
             outbag.write(topic, msg, t, raw=True, connection_header=conn_header)
-            
-            total_bytes += len(serialized_bytes) 
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
            
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 41]
```diff
@@ -787,12 +789,12 @@ def change_compression_op(inbag, outbag, compression, quiet):
...
+            total_bytes += len(serialized_bytes)
             meter.step(total_bytes)
-        
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
        
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 42]
```diff
@@ -879,7 +881,7 @@ def __init__(self):
     def add_cmd(self, name, function, description):
         self[name] = function
         self._description[name] = description
-        
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
        
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 43]
```diff
@@ -914,7 +916,7 @@ def __init__(self, path, bytes_total, refresh_rate=1.0):
         self.path           = path
         self.bytes_total    = bytes_total
         self.refresh_rate   = refresh_rate
-        
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
        
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 44]
```diff
@@ -926,7 +928,7 @@ def __init__(self, path, bytes_total, refresh_rate=1.0):
     def step(self, bytes_read, force_update=False):
         self.bytes_read = bytes_read
         self.elapsed    = time.time() - self.start_time
-        
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
        
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 45]
```diff
@@ -939,7 +941,7 @@ def _update_progress(self):
 
         bytes_read_str  = self._human_readable_size(float(self.bytes_read))
         bytes_total_str = self._human_readable_size(float(self.bytes_total))
-        
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
        
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 46]
```diff
@@ -960,14 +962,14 @@ def _update_progress(self):
 
         print('\r', progress, end='')
         sys.stdout.flush()
-        
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
        
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 47]
```diff
@@ -960,14 +962,14 @@ def _update_progress(self):
...
     def _human_readable_size(self, size):
         multiple = 1024.0
         for suffix in ['KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB']:
             size /= multiple
             if size < multiple:
                 return '%.1f %s' % (size, suffix)
-    
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
    
```

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 48]
```diff
@@ -992,7 +994,7 @@ def terminal_width():
                 pass
         if width <= 0:
             width = 80
-    
+
```
Comment (tappan-at-git) [inline tools/rosbag/src/rosbag/rosbag_main.py]:
```suggestion
    
```

**Review by tappan-at-git — COMMENTED** [review]
(Marking all automatic whitespace changes)

**Review by tappan-at-git — COMMENTED** [review]
(Marking all automatic whitespace changes)

**Comment by tappan-at-git** [comment]
Welp. So much for bypassing my editor's whitespace plugins by using github. It won't let me batch those comments for. Some reason.

**Review by paulbovbel — APPROVED** [review]
LGTM, given discussion offline about pushing this to an upstream PR after removing whitespace changes.

**Review by abencz — APPROVED** [review]
LGTM 👍

**Pull request merged**